onnxruntime/core/optimizer/insert_cast_transformer.cc

```diff
@@ -95,30 +119,33 @@ class RemoveDuplicateCastTransformer : public GraphTransformer {
       bool removed = false;
       if (node.OpType() == "Cast") {
         std::vector<std::reference_wrapper<Node>> nodes_to_remove;
+        std::vector<std::reference_wrapper<Node>> cast_nodes_to_keep;
 
         // if cast's next node is also cast and next cast's output type equal to cast's input type
         // remove those two cast.
-        // boolean is an exception case for this optimization
+        // Any Cast that changing the value is an exception case for this optimization.
         auto src_type = node.InputDefs()[0]->Type();
         auto dst_type = node.OutputDefs()[0]->Type();
-        if (*src_type == "tensor(bool)" || *dst_type == "tensor(bool)")
+        TypeGroup src_type_group = GetTypeGroup(src_type);
+        TypeGroup dst_type_group = GetTypeGroup(dst_type);
+        if (src_type_group == Unknown || dst_type_group == Unknown || src_type_group > dst_type_group) {
```

Comment:
Not too sure about this "src_type_group > dst_type_group" condition... 

If we have following chain of casts "float-> int -> bool", can this is be simplified to "float->bool"

Reply:
Let's add a comment somewhere explaining the rules here

- It's **not ok** for (high precision -> low precision -> high precision) to skip the middle step, since there is actual loss of precision
- It's **ok** for (low precision -> high precision ->low precision) to skip the middle step, since there is no actual loss of precision
- it's **ok** for (high precision -> low precision -> lower precision) to skip the middle step.

Reply:
Previously I just wanted to fix the bug that the original code just handled bool base in the first case, the code didn't look into the second cast node so I didn't either. You are right that (high precision -> low precision -> lower precision) is also OK for this optimization. I just updated the rules to cover this case.